find_broken_links.py

```diff
@@ -0,0 +1,118 @@
+import re
+import socket
+import sys
+import urllib.request
+from glob import iglob
+from typing import Iterator, TextIO, Tuple
+from urllib.error import HTTPError, URLError
+
+from dataclasses import dataclass
+
+REQUEST_HEADERS = {'User-Agent': 'https://github.com/RTBHOUSE/here-be-pythons'}
+TIMEOUT_SECONDS = 5
+URL_REGEX = re.compile(
+    r'(?:http|ftp|https)://'                    # protocol
+    r'(?:[\w_-]+(?:(?:\.[\w_-]+)+))'            # domain
+    r'(?:[\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])?'  # path
+)  # yapf: disable
+
+
+def find_broken_links(files_glob_pattern='docs/source/**/*.rst') -> None:
+    counters = Counters()
+
+    for file in _yield_files(files_glob_pattern):
+        _process_urls_from_file(file, counters)
+
+    _print_report(counters)
+
+    if counters.broken_links_num == 0:
+        sys.exit(0)
+    else:
+        sys.exit(1)
+
+
+@dataclass
+class Link:
+    url: str
+    health: str = None
+    status_code: int = None
+
+
+@dataclass
+class Counters:
+    ok_links_num: int = 0
+    broken_links_num: int = 0
+    timeout_links_num: int = 0
+
+
+def _yield_files(files_glob_pattern: str) -> Iterator[TextIO]:
+    files_paths = iglob(files_glob_pattern)
+    for file_path in files_paths:
+        with open(file_path) as file:
+            yield file
+
+
+def _process_urls_from_file(file: TextIO, counters: Counters) -> None:
+    for link, line_num in _yield_links_from_file(file):
+        _update_link_health(link, counters)
+        if link.health == 'broken':
```

Comment:
Why link need `health` field? Maybe you can just return it from method line above and check?

Reply:
Been there, seen that!

It turned out after numerous iterations/refactors that using a simple `Link` object:
+ reduces code clutter (there are less values returned by functions, less temporary variables to handle, etc.),
+ makes it more semantic-driven (i.e. it is easier to read and reason about the code).